### ISSUE TITLE
Uns categorical colour maping as a dictionary 

### ISSUE THREAD

**Issue by Hrovatin** [scverse/scanpy#1340]
- [ x] Additional function parameters / changed functionality / changed defaults?
- [ ] New analysis tool: A simple analysis tool you have been using and are missing in `sc.tools`?
- [ ] New plotting function: A kind of plot you would like to seein `sc.pl`?
- [ ] External tools: Do you know an existing package that should go into `sc.external.*`?
- [ ] Other?

When determining colours of categorical variables in uns they are based on alphabetical order (if I am not mistaken) - being represented just as an ordered list. Thus it is a bit inconvenient to change colours, especially if categories change during the analysis - the whole order changes and the mapping breaks. Would it be possible to use a dictionary of colours as values and categories as keys (with a default for any categories missing colours)?

**Comment by LuckyMD** [comment]
I feel like this discussion was already had and it was agreed that this should be possible and someone implemented it, no? @fidelram @gokceneraslan @ivirshup ? I found sth about using a dictionary structure in #1134 but it wasn't addressed.

Regarding the ordering of colours in `.uns['myVar_colors']`, this is ordered by `.obs['myVar'].cat.categories`.

**Comment by ivirshup** [comment]
It's definitely been discussed and is something I would like to do, but it has not been implemented. I think the main blocker here is that it would be a lot of effort to go through every plotting function and make it work with that structure.

A good step towards this would be normalizing access to colors through a single function.

In general, I think the colors could be stored as something like:

```
.uns["colors"][dim][key] = {cat1: color1, ...}
``` 

though I'm open to other solutions.

**Comment by LuckyMD** [comment]
I think your suggestion sounds good, but wouldn't it be easier if the syntax with `adata.obs['myVar_colors']` was kept so that less needs to be changed?

**Comment by ivirshup** [comment]
I think we should allow for categorical colors along either axis, and right now it's becomes ambiguous. A good example of an annotation that can apply to both observations and variables is `species`.

I'd like to shift to a nested model to limit the amount of reserved keys in `.uns`. It reduces that chance of unintentional naming collisions.

As for the amount of things that would need to change, a lot has to change anyways. Hardly any code that works with the current setup will work with mappings (`len` is all I can think of). If we're already making a breaking change, might as well take advantage and future proof it a bit.

**Comment by LuckyMD** [comment]
Okay, so the plan would be a breaking change. So scanpy 1.7 then?

**Comment by ivirshup** [comment]
Depends on how we're feeling about semantic versioning. 2.0 for a complete switch, but we could internally switch over with a compatibility layer and deprecation warnings anytime before then.

**Comment by LuckyMD** [comment]
I like the idea of a compatability layer. @Hrovatin is this something you would like to contribute to as well?

**Comment by fidelram** [comment]
I agree with @ivirshup that the best would be to have a single function that returns the colors. This also facilitates documentation. The actual implementation afterwards could be modified without much effort.
